python/ray/client_builder.py

```diff
@@ -100,19 +110,37 @@ def connect(self) -> ClientContext:
                 includes the server's version of Python & Ray as well as the
                 dashboard_url.
         """
+        old_ray_cxt = None
+        if ray.util.client.ray.is_connected(
+        ) and self._allow_multiple_connections:
+            raise RuntimeError("It's in single-client mode. We can't create"
+                               " new client in multi-client mode")
+
+        if ray.util.client.num_connected_contexts(
+        ) != 0 and not self._allow_multiple_connections:
+            raise RuntimeError("Multiple clients are connected. Additional "
+                               "connections must be created with"
+                               " `ray.init(allow_multiple=True)`")
+
+        if self._allow_multiple_connections:
+            old_ray_cxt = ray.util.client.ray.set_context(None)
         client_info_dict = ray.util.client_connect.connect(
             self.address,
             job_config=self._job_config,
             ray_init_kwargs=self._remote_init_kwargs)
         dashboard_url = ray.get(
             ray.remote(ray.worker.get_dashboard_url).remote())
-        return ClientContext(
+        cxt = ClientContext(
             dashboard_url=dashboard_url,
             python_version=client_info_dict["python_version"],
             ray_version=client_info_dict["ray_version"],
             ray_commit=client_info_dict["ray_commit"],
             protocol_version=client_info_dict["protocol_version"],
-            _num_clients=client_info_dict["num_clients"])
+            _num_clients=client_info_dict["num_clients"],
+            _context_to_restore=ray.util.client.ray.get_context())
```

Comment:
Can we just instantiate a new _ClientContext() here?

Reply:
The new client context is managed by ray.util.client.ray, so this actually is the new context. It's not reusing anything.

My method here is trying to hide everything from the outside, so all the creation kind of thing will be done in ray.util.client.ray automatically and the outside world just need to set up the context.

If the context is not set up, it'll use the global one (the one created without allow_multiple=True).

Reply:
Ahhh, that makes sense :)